src/controller/CHIPDiscovery.cpp

```diff
@@ -0,0 +1,336 @@
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include <stdarg.h>
+
+#include <cstdint>
+#include <string>
+
+#include <controller/CHIPDiscovery.h>
+#include <lib/mdns/minimal/core/Constants.h>
+
+namespace chip {
+namespace Controller {
+
+namespace {
+std::string ToString(const mdns::Minimal::BytesRange & bytes)
+{
+    std::string str;
+    str.reserve(bytes.Size() + 1);
+    str.assign(reinterpret_cast<const char *>(bytes.Start()), bytes.Size());
+    return str;
+}
+
+} // namespace
+
+void TxtRecordDelegateImpl::OnRecord(const mdns::Minimal::BytesRange & name, const mdns::Minimal::BytesRange & value)
+{
+    if (dnsSdInfo == nullptr)
+    {
+        return;
+    }
+    // Names are just 1 letter.
+    if (name.Size() < 1)
+    {
+        return;
+    }
+    std::string key = ToString(name);
+    std::string val = ToString(value);
+    if (key == "D")
+    {
+        // For some reason there is no standard string -> short converstion.
+        unsigned long temp = std::stoul(val);
+        if (temp < UINT16_MAX)
+        {
+            dnsSdInfo->longDiscriminator = static_cast<uint16_t>(temp);
+        }
+    }
+    else if (key == "VP")
+    {
+        // Fist value is the vendor id, second (after the +) is the product.
+        size_t plusSignIdx = val.find('+');
+        unsigned long temp = std::stoul(val.substr(0, plusSignIdx));
+        if (temp < UINT16_MAX)
+        {
+            dnsSdInfo->vendorId = static_cast<uint16_t>(temp);
+        }
+        if (plusSignIdx == std::string::npos || (plusSignIdx + 1) >= val.length())
+        {
+            return;
+        }
+        temp = std::stoul(val.substr(plusSignIdx + 1, std::string::npos));
+        if (temp < UINT16_MAX)
+        {
+            dnsSdInfo->productId = static_cast<uint16_t>(temp);
+        }
+    }
+}
+
+void ParserDelegateImpl::OnHeader(mdns::Minimal::ConstHeaderRef & header)
+{
+    // Controller doesn't care about headers.
+}
+void ParserDelegateImpl::OnQuery(const mdns::Minimal::QueryData & data)
+{
+    // Controller doesn't care about queries.
+}
+
+int ParserDelegateImpl::FindDnsSdByHostName(const char * hostName) const
+{
+    for (int i = 0; i < kNumDnsSdEntries; ++i)
+    {
+        if (dnsSdInfo[i].IsInstance(hostName))
+        {
+            return i;
+        }
+    }
+    return -1;
+}
+
+int ParserDelegateImpl::FindDnsSdByInstanceName(const char * instanceName) const
+{
+    for (int i = 0; i < kNumDnsSdEntries; ++i)
+    {
+        if (dnsSdInfo[i].IsInstance(instanceName))
+        {
+            return i;
+        }
+    }
+    return -1;
+}
+
+DiscoveryType GetDiscoveryTypeFromQname(mdns::Minimal::SerializedQNameIterator it)
+{
+    while (it.Next())
+    {
+        if (strncmp(it.Value(), "_chipc", strlen("_chipc")) == 0)
```

Comment:
Can you add commissioner discovery (chipd) as well? Most of the parsing code can be shared since the data types are the same.